src/webgpu/api/operation/memory_sync/buffer.spec.ts

```diff
@@ -0,0 +1,257 @@
+  // Otherwise, write 2 into the buffer.
+  createComputePipeline(isFirstWrite: boolean): GPUComputePipeline {
+    const value: string = isFirstWrite ? '1' : '2';
+    const wgslCompute = `
+      type Data = [[block]] struct {
+        [[offset 0]] a : i32;
+      };
+
+      [[binding 0, set 0]] var<storage_buffer> data : Data;
+      fn main() -> void {
+	data.a = ${value};
+        return;
+      }
+
+      entry_point compute = main;
+    `;
+
+    return this.device.createComputePipeline({
+      computeStage: {
+        module: this.device.createShaderModule({
+          code: wgslCompute,
+        }),
+        entryPoint: 'main',
+      },
+    });
+  }
+
+  // Create a render pipeline and write data into storage buffer at fragment stage.
+  // If the pipeline is used in the first write, then write 1 into the buffer.
+  // Otherwise, write 2 into the buffer.
+  createRenderPipeline(isFirstWrite: boolean): GPURenderPipeline {
+    const wgslVertex = `
+    [[builtin position]] var<out> Position : vec4<f32>;
+      fn vert_main() -> void {
+        Position = vec4<f32>(0.5, 0.5, 0.0, 1.0);
+        return;
+      }
+
+      entry_point vertex = vert_main;
+    `;
+
+    const value: string = isFirstWrite ? '1' : '2';
+    const wgslFragment = `
+    [[location 0]] var<out> outColor : vec4<f32>;
+      type Data = [[block]] struct {
+        [[offset 0]] a : i32;
+      };
+
+      [[binding 0, set 0]] var<storage_buffer> data : Data;
+      fn frag_main() -> void {
+	data.a = ${value};
+	outColor = vec4<f32>(1.0, 0.0, 0.0, 1.0);
+        return;
+      }
+
+      entry_point fragment = frag_main;
+    `;
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'frag_main',
+      },
+      primitiveTopology: 'point-list',
+      colorStates: [{ format: 'rgba8unorm' }],
+    });
+  }
+
+  beginSimpleRenderPass(encoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const view = this.device
+      .createTexture({
+        size: { width: 4, height: 4, depth: 1 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+  }
+
+  // Create render pass or compute pass, set pipeline and bind groups and write data into storage
+  // buffer. If the pass is render pass, it may use bundle. Note that we write 1 into the buffer
+  // if the pass is used in the first write, and we write 2 into the buffer otherwise.
+  createPass(
```

Comment:
The chain of ifs here, in createCommandBuffer and in the test is a bit difficult to understand. What about the following: we could have a "context" structure that can contain all 4 types of encoders. Each of the usage helper methods would call it with "ensureComputeEncoder()" similarly to what we do in Dawn with the Metal EncodingContext. This context structure would have the logic to end encoders or start new ones if needed.

Then the top level of the test can become:

```
context = new Context(device, queue);
value = 1;
operationType[firstOperation].performWrite(context, value);

// Either as a bunch of ifs, or as a helper that hides that logic in the context
if (differentBundle) {
    context.InvalidateBundle();
}
if (differentPass) {
    context.InvalidatePass();
}

value ++;
operationType[firstOperation].performWrite(context, value);
```

Turns out that with this structure the read tests become exactly the same thing!

Reply:
Thanks for your suggestion, Corentin. I actually wrote tests for write operations like copy, etc as you mentioned above. It turned out that we can use helper functions here. I updated code to be aligned with the pull request followed. And the new one is easy to extend to more write operations like t2b copy, b2b copy, writeBuffer. Now that we have the helper functions, it is much easier to understand and it is extensible too. Can you take a look at this solution which is already there?